language/move-core/types/src/resolver.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    account_address::AccountAddress,
+    language_storage::{ModuleId, StructTag},
+};
+use std::fmt::Debug;
+
+/// Traits for resolving Move modules and resources from persistent storage
+
+/// A persistent storage backend that can resolve modules by address + name.
+/// Storage backends should return
+///   - Ok(Some(..)) if the data exists
+///   - Ok(None)     if the data does not exist
+///   - Err(..)      only when something really wrong happens, for example
+///                    - invariants are broken and observable from the storage side
+///                      (this is not currently possible as ModuleId and StructTag
+///                       are always structurally valid)
+///                    - storage encounters internal error
+pub trait ModuleResolver {
+    type Error: Debug;
+
+    fn get_module(&self, id: &ModuleId) -> Result<Option<Vec<u8>>, Self::Error>;
+}
+
+/// A persistent storage backend that can resolve resources by address + type
+/// Storage backends should return
+///   - Ok(Some(..)) if the data exists
+///   - Ok(None)     if the data does not exist
+///   - Err(..)      only when something really wrong happens, for example
+///                    - invariants are broken and observable from the storage side
+///                      (this is not currently possible as ModuleId and StructTag
+///                       are always structurally valid)
+///                    - storage encounters internal error
+pub trait ResourceResolver {
+    type Error: Debug;
+
+    fn get_resource(
+        &self,
+        address: &AccountAddress,
+        typ: &StructTag,
+    ) -> Result<Option<Vec<u8>>, Self::Error>;
+}
+
+/// A persistent storage implementation that can resolve both resources and modules
+pub trait MoveResolver<Error: Debug>:
+    ModuleResolver<Error = Error> + ResourceResolver<Error = Error>
+{
```

Comment:
Sorry about going on this issue back and force... I think I now see where my confusion is coming from. I was wondering why are we having `Error` as type input instead of doing the associate type definition as we did for `ResourceResolver` and `ModuleResolver`?

Reply:
Good point, the suggestion you sent me looks much cleaner/more consistent than what I have.